text/0010-bft-signatures.md

```diff
@@ -0,0 +1,291 @@
+
+While forks of Fabric with a Byzantine Fault Tolerant ordering service [exist](https://github.com/SmartBFT-Go/fabric/), they have not been incorporated into the official Hyperledger Fabric project.
+
+This RFC aims to:
+
+- Give an overview of how denial of service resistant block retrieval and block signature validation is done in SmartBFT.
+- Define how denial of service resistant block retrieval should be conducted in Hyperledger Fabric.
+- Define how Byzantine block signature validation should be constructed and encoded in Hyperledger Fabric.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, Hyperledger Fabric only supports ordering service types that are Crash Fault Tolerant (CFT), not Byzantine Fault Tolerant (BFT).
+
+There are two main attacks that are considered outside of the crash fault tolerant thread model, but must be protected against in the Byzantine fault tolerant setting:
+
+**Block withholding**: A Byzantine orderer might be reluctant to send newly formed blocks to peers that connect to it, either because of malicious intent or to save egress network bandwidth charges from its hosting provider.
+ Fabric architecture relies on peers receiving newly formed blocks as fast as possible, because otherwise client transaction submission times out and smart contract simulation results in stale reads.
+
+**Block forgery**: A Byzantine orderer might try to forge a block and send it to peers or other ordering nodes that replicate from it, in order to alter the smart contract execution on peers by subterfuge.
+ One way of doing this, is to omit a transaction from a block, or re-order transactions within a block.
+ 
+To circumvent block forgery attacks, the following invariant should be preserved: *A set of signatures on a block should only be considered valid if it is impossible for the parties running the protocol to construct a set of signatures on a different block that is also considered valid*.
+
+The aforementioned invariant can be implemented in different ways, such as requiring a **threshold of signatures** from different orderers on a block, or requiring a single **threshold signature** on a block.
+The threshold itself can vary depending on the consensus algorithm and on the stage in it that the signatures is produced: Signing a block post consensus requires signatures from only `f+1` orderers (at the expense of latency), while signing during consensus requires signatures from `2f+1` orderers (at the expense of size and computation).
+
+
+# Naive approaches for block replication and block signature verification
+
+### Block replication
+
+In a Byzantine setting, there is a fundamental assumption that there is an upper bound on the number of malicious or failed nodes
+ among all the nodes that run the protocol, and that upper bound is denoted as `f`.
+ In BFT, the ordering service is comprised of `n>3f` for some `f>0`, so we have up to `f` nodes which are either offline, unreachable or malicious,
```

Comment:
Is it correct that `n` has to be odd?

Reply:
`n` is odd when `f` is even and vice versa.

Reply:
> `n` is odd when `f` is even and vice versa.

if one sets _f_ possibly high, which perhaps makes the most sense. However, in principle, one can, say, set _f=1_ and _n=5_.

Reply:
> _f=1_ and _n=5_

In fact, this seems the only possible configuration if there are 5 independent stakeholders, each wanting to operate a PBFT node.

Reply:
Indeed, I am often simplifying by assuming `n=3f+1`. 

In reality the number of messages you collect instead of `2f+1` is from a "quorum" of nodes, and it's defined in a way that every two quorums have at least a single correct node.